src/include/storage/index/hash_index.h

```diff
@@ -0,0 +1,272 @@
+                   "Called index delete on a TupleSlot that has a conflict with this txn or is still visible.");
+
+    // Register a deferred action for the GC with txn manager. See base function comment.
+    auto *const txn_manager = txn->GetTransactionManager();
+    txn->RegisterCommitAction([=]() {
+      txn_manager->DeferAction([=]() {
+        auto abort_key_found_fn = [location](ValueType &value) -> bool {
+          if (std::holds_alternative<TupleSlot>(value)) {
+            // It's just a TupleSlot, functor should return true for uprase to erase it
+            return true;
+          }
+          auto &value_map = std::get<ValueMap>(value);
+          if (value_map.size() == 1) {
+            // ValueMap only has 1 element, functor should return true for uprase to erase it
+            TERRIER_ASSERT(value_map.count(location) == 1, "location must be the only value in the ValueMap.");
+            return true;
+          }
+          // ValueMap contains multiple elements, erase the element for location
+          const auto UNUSED_ATTRIBUTE erase_result = value_map.erase(location);
+          TERRIER_ASSERT(erase_result == 1, "Erasing from the ValueMap should not fail.");
+          return false;
+        };
+
+        const bool UNUSED_ATTRIBUTE uprase_result = hash_map_->uprase_fn(index_key, abort_key_found_fn);
+        TERRIER_ASSERT(!uprase_result, "This operation should NOT insert a new key into the cuckoohash_map.");
+      });
+    });
+  }
+
+  void ScanKey(const transaction::TransactionContext &txn, const ProjectedRow &key,
+               std::vector<TupleSlot> *value_list) final {
+    TERRIER_ASSERT(value_list->empty(), "Result set should begin empty.");
+
+    // Build search key
+    KeyType index_key;
+    index_key.SetFromProjectedRow(key, metadata_);
+
+    auto key_found_fn = [value_list, &txn](const ValueType &value) -> void {
+      if (std::holds_alternative<TupleSlot>(value)) {
+        const auto existing_location = std::get<TupleSlot>(value);
+        if (IsVisible(txn, existing_location)) value_list->emplace_back(existing_location);
+      } else {
+        const auto &value_map = std::get<ValueMap>(value);
+
+        for (const auto i : value_map) {
+          if (IsVisible(txn, i)) value_list->emplace_back(i);
+        }
+      }
+    };
+
+    const bool UNUSED_ATTRIBUTE find_result = hash_map_->find_fn(index_key, key_found_fn);
+
+    TERRIER_ASSERT(GetConstraintType() == ConstraintType::DEFAULT ||
+                       (GetConstraintType() == ConstraintType::UNIQUE && value_list->size() <= 1),
+                   "Invalid number of results for unique index.");
+  }
+
+  void ScanAscending(const transaction::TransactionContext &txn, const ProjectedRow &low_key,
+                     const ProjectedRow &high_key, std::vector<TupleSlot> *value_list) final {
+    TERRIER_ASSERT(false, "Range scans not supported on HashIndex.");
```

Comment:
These (and the following asserts) may need to be exceptions or something stronger.  Can we guarantee that a user cannot force us down this path?  I would like to get there but there is a lot of logic we haven't written to ensure this.